yt/yt/core/logging/log_writer_detail.cpp

```diff
@@ -73,50 +73,128 @@ i64 TRateLimitCounter::GetAndResetLastSkippedEventsCount()
+{ }
+
+void TSegmentSizeReporter::IncrementSegmentSize(i64 size)
+{
+    CurrentSegmentSizeGauge_.Update(CurrentSegmentSize_.fetch_add(size, std::memory_order::relaxed) + size);
+}
+
+void TSegmentSizeReporter::ResetSegmentSize(i64 size)
+{
+    CurrentSegmentSize_.store(size, std::memory_order::relaxed);
+    CurrentSegmentSizeGauge_.Update(size);
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+TRateLimitingLogWriterBase::TRateLimitingLogWriterBase(TCallback<i64(const TLogEvent&)> writeImpl, TString name, TLogWriterConfigPtr config)
+    : WriteImpl_(writeImpl)
     , Name_(std::move(name))
+    , Config_(std::move(config))
     , RateLimit_(
         std::nullopt,
         {},
         TProfiler{"/logging"}.WithSparse().WithTag("writer", Name_).Counter("/events_skipped_by_global_limit"))
-    , CurrentSegmentSizeGauge_(
-        TProfiler{"/logging"}.WithSparse().WithTag("writer", Name_).Gauge("/current_segment_size"))
 { }
 
-void TStreamLogWriterBase::Write(const TLogEvent& event)
+void TRateLimitingLogWriterBase::Write(const TLogEvent& event)
 {
-    auto* stream = GetOutputStream();
-    if (!stream) {
-        return;
+    auto* categoryRateLimit = GetCategoryRateLimitCounter(event.Category->Name);
+    if (RateLimit_.IsIntervalPassed()) {
+        auto eventsSkipped = RateLimit_.GetAndResetLastSkippedEventsCount();
+        WriteLogSkippedEvent(eventsSkipped, Name_);
+    }
+    if (categoryRateLimit->IsIntervalPassed()) {
+        auto eventsSkipped = categoryRateLimit->GetAndResetLastSkippedEventsCount();
+        WriteLogSkippedEvent(eventsSkipped, event.Category->Name);
     }
 
-    try {
-        auto* categoryRateLimit = GetCategoryRateLimitCounter(event.Category->Name);
-        if (RateLimit_.IsIntervalPassed()) {
-            auto eventsSkipped = RateLimit_.GetAndResetLastSkippedEventsCount();
-            if (eventsSkipped > 0) {
-                Formatter_->WriteLogSkippedEvent(stream, eventsSkipped, Name_);
-            }
-        }
-        if (categoryRateLimit->IsIntervalPassed()) {
-            auto eventsSkipped = categoryRateLimit->GetAndResetLastSkippedEventsCount();
-            if (eventsSkipped > 0) {
-                Formatter_->WriteLogSkippedEvent(stream, eventsSkipped, event.Category->Name);
-            }
-        }
-        if (!RateLimit_.IsLimitReached() && !categoryRateLimit->IsLimitReached()) {
-            auto bytesWritten = Formatter_->WriteFormatted(stream, event);
-            CurrentSegmentSize_ += bytesWritten;
-            CurrentSegmentSizeGauge_.Update(CurrentSegmentSize_);
+    // We do not do this in the constructor since it happens to be called during TLogManager construction itself.
+    if (!SystemCategory_) {
+        SystemCategory_ = GetDefaultLogManager()->GetCategory(SystemLoggingCategoryName);
+    }
+    bool isSystemEvent = event.Category == SystemCategory_;
```

Comment:
It is new, it seemed reasonable from a general perspective when I was writing the code.

Now that I think about it, I can't say it makes too much of a difference: log writers that would receive system category events won't actually have any limits configured, since they are handled by a designated error log writer anyway. Meanwhile, log skipped system events will get written in any case, since they are not looped back into the rate limiting logic. 

I am fine with reverting this part, if you feel like it would improve readability.

Reply:
I'm not really worried about readability here, I just couldn't verify that this is 100% safe with a quick glance (and I couldn't allocate more time to think about it, given that it's not essential for the PR's primary goal).

Not really a strong opinion, but reverting it would make the code a bit tidier, yes.

Reply:
Reverted this part.